python/lsst/daf/butler/registry/wildcards.py

```diff
@@ -467,51 +358,44 @@ def __repr__(self) -> str:
         return f"CollectionSearch({self.__root__!r})"
 
 
-class CollectionQuery:
-    """An unordered query for collections and dataset type restrictions.
+@dataclasses.dataclass(frozen=True)
+class CollectionWildcard:
+    """A validated wildcard for collection names
 
-    The `fromExpression` method should almost always be used to construct
+    The `from_expression` method should almost always be used to construct
     instances, as the regular constructor performs no checking of inputs (and
     that can lead to confusing error messages downstream).
 
-    Parameters
-    ----------
-    search : `CollectionSearch` or `...`
-        An object representing an ordered search for explicitly-named
-        collections (to be interpreted here as unordered), or the special
-        value `...` indicating all collections.  `...` must be accompanied
-        by ``patterns=None``.
-    patterns : `tuple` of `re.Pattern`
-        Regular expression patterns to match against collection names.
-    universe : `DimensionUniverse`
-        Object managing all dimensions.
-
     Notes
     -----
-    A `CollectionQuery` is used to find all matching datasets in any number
-    of collections, or to find collections themselves.
-
-    `CollectionQuery` is expected to be rarely used outside of `Registry`
+    `CollectionWildcard` is expected to be rarely used outside of `Registry`
     (which uses it to back several of its "query" methods that take general
-    expressions for collections), but it may occassionally be useful outside
+    expressions for collections), but it may occasionally be useful outside
     `Registry` as a way to preprocess expressions that contain single-pass
     iterators into a form that can be used to call those `Registry` methods
     multiple times.
     """
 
-    def __init__(
-        self,
-        search: CollectionSearch | EllipsisType = Ellipsis,
-        patterns: tuple[re.Pattern, ...] = (),
-    ):
-        self._search = search
-        self._patterns = patterns
+    strings: tuple[str, ...] = ()
+    """An an ordered list of explicitly-named collections. (`tuple` [ `str` ]).
+    """
 
-    __slots__ = ("_search", "_patterns")
+    patterns: tuple[re.Pattern, ...] | EllipsisType = Ellipsis
+    """Regular expression patterns to match against collection names, or the
+    special value ``...`` indicating all collections.
+
+    `...` must be accompanied by ``strings=()``.
+    """
+
+    def __post_init__(self) -> None:
+        if self.patterns is Ellipsis and self.strings:
+            raise ValueError(
+                f"Collection wildcard matches any string, but still has explicit strings {self.strings}."
+            )
```

Comment:
Is there a chance that we can get `.*` pattern here, which would also match anything?